### ISSUE TITLE
fix(model): enforce sampled-row projection floor on lm_head to prevent chunked-prefill OOM

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12080]
# fix(model): enforce sampled-row projection floor on lm_head to prevent chunked-prefill OOM

```routing
lane: internal/model
paths: ["internal/model/batch_prefill.go", "internal/compute/lm_head_test.go"]
expected_steps: 4
```

### Parent context
Parent: #9204 (epic(model): QWEN38-FLASH-NEXT fak-native support)

### Current state
In `fak` today, model execution during chunked prefill can evaluate linear vocabulary projections (`lm_head`) across wide sequence chunk windows ($M = 8,192$). On models with large vocabularies ($V = 248,320$, such as Qwen 3.5 / Qwen 3.8 / DeepSeek), materializing the intermediate logits tensor across the full sequence chunk consumes $8,192 \times 248,320 \times 2\text{ bytes} \approx 3.79\text{ GiB}$ of transient GPU memory. On 32GB GPUs (RTX 5090D, RTX 4090), this unexpected multi-gigabyte memory spike causes an immediate out-of-memory crash on the very first prefill request.

### Why this is next
Autoregressive language models only sample next tokens from the terminal token position of each active request. Computing and materializing vocabulary logits for internal prompt tokens that are never sampled is completely wasted compute ($10.4\text{ TFLOPs}$) and creates catastrophic memory volatility. Slicing the hidden states before the projection guarantees zero OOM risk during large chunked prefills.

### Priority context
Working path: model -> prefill -> lm_head -> sample. Current blocker: 3.79 GiB logits allocation spike on 8k chunks. Unblocks: 32k+ chunked prefill on 24GB/32GB consumer hardware. Not polish: crash prevention.

### Working spine
Enforce a sampled-row projection floor on `lm_head` forward dispatch that slices input hidden states to sampled query positions before vocabulary GEMM.

## Value
- Centrality: Core
- P1: advanced - eliminates 3.79 GiB prefill transient logits spike on 8k chunk prefill
- P2: advanced - saves 10.4 TFLOPs of unnecessary vocabulary projections per chunk
- P3: preserved - logit numerical outputs bit-match full projection on sampled rows
- P4: preserved - unconstrained full-sequence projection remains available for training/perplexity

## Named external inspiration
- **Source:** FreeToken (`FlashML-org/FreeToken@af71ba43206e124f5ff6419b47ee36c6e9981078`), PR #342 (`qwen3_5_moe: run lm_head on sampled rows only (fixes 32 GB first-prefill OOM)`), `python/freetoken/models/qwen3_5_moe/model.py:112-121`.
- **Mechanism:** Slices the hidden state tensor prior to the linear projection: `self.lm_head.forward(output[: ctx.batch.size])`. Reduces the transient logits buffer from $3.79\text{ GiB}$ to $496\text{ KB}$ ($7,800\times$ reduction at $B=1$), eliminating the first-prefill OOM entirely and saving $10.4\text{ TFLOPs}$ per 8k chunk.
- **License:** Apache-2.0 (ADAPT into fak-native compute/model layers).

## Core through-line
Deliver a fail-closed sampled-row constraint on `lm_head` operators in `internal/model/` and `internal/compute/` that:
1. Inspects batch execution phase and active query sample indices before issuing the vocabulary GEMM.
2. Restricts the GEMM operand dimension $M$ to `batch.size` (or designated sample indices) during prefill.
3. Completely bypasses logit generation for non-terminal tokens unless explicit evaluation is requested (e.g. perplexity / training loss).
4. Verifies zero memory regressions and exact logit parity on sampled positions.

## Gold-plating boundary
Strictly excluded from this work unit:
- Vocabulary pruning or dynamic candidate logit filtering.
- Speculative verification head micro-batching.
- Logits processor modifications.

### Done condition / witness
Done condition: `internal/model/batch_prefill.go` slices input activations to sampled token indices before invoking vocabulary projection, capping transient logits to sampled positions.
Witness: `go -C ../fak test -v -race ./internal/model -run TestLMHeadSampledRowFloor -count=1`

### Acceptance gate
go -C ../fak test -v -race ./internal/model -run TestLMHeadSampledRowFloor -count=1

## Verifiable Witness
```bash
go -C ../fak test -v -race ./internal/model -run TestLMHeadSampledRowFloor -count=1
```

## Witness
```bash
go -C ../fak test -v -race ./internal/model -run TestLMHeadSampledRowFloor -count=1
```

### File:Line Seam
`internal/model/batch_prefill.go:168`

### Blast Radius & Affected Lanes
- **Primary Lane:** `internal/model`
- **Affected Packages:** `internal/model`, `internal/compute`
- **Unaffected Lanes:** decoupled; trunk remains 100% green

### Quarantined Fallback Mechanism
If specialized slicing is bypassed, execution falls back to full-tensor projection with a memory warning.

### Closure binding
Resolving commit cites this issue.

## Scoped Acceptance Criteria
- [ ] 1. `internal/model/batch_prefill.go` projects only sampled query rows to vocabulary logits during chunked prefill.
- [ ] 2. Memory allocation for 8,192-token chunk on 248k vocabulary does not exceed 10 MB (vs 3.79 GB previously).
- [ ] 3. Sampled logit outputs bit-match full-projection logits exactly.
- [ ] 4. Verifiable witness passes: `go -C ../fak test -v -race ./internal/model -run TestLMHeadSampledRowFloor -count=1`.

## Likely files
- `internal/model/batch_prefill.go`
- `internal/compute/lm_head_test.go`

## Lane
internal/model

## Work estimate
Estimate: 3 points. Uncertainty: operator-supplied estimate; revise through the parent denominator when evidence changes.

## Overall completion contribution
Contribution: 3/20 points.

## Completion standard
production

## Target operating envelope
acceptance pass rate: = 100 percent

## Witnessed operating envelope
acceptance pass rate: = 100 percent

**Comment by anthony-chaudhary** [comment]
Claimed for a bounded goal wave. Scope: sampled-row lm_head projection floor with fail-before/pass-after package witness; no vocabulary pruning or speculative-head expansion.

**Comment by anthony-chaudhary** [comment]
Resolved by 9fbe0fa209af. Rectangular F32/Q8 prefill now uses an explicit sampled-row lm_head selector, rejects invalid/missing plans, preserves the full Forward evaluation path, and binds the 8192x248320 fp16 geometry below 10 MiB. Witness: go test -v -race ./internal/model -run '^TestLMHeadSampledRowFloor$' -count=1. CUDA/Vulkan sequence-prefill was also verified already terminal-row-only.

**Comment by anthony-chaudhary** [comment]
Reopened after independent review of 9fbe0fa209af. Remaining P2s: zero-value mode currently selects full projection despite the explicit/fail-closed contract; the test does not bind production F32/Q8 lm_head M; and Q8 rectangular prefill is not exercised. Follow-up will make zero unspecified, derive product row counts from the selected panel, and add an independent quantized-path parity witness.

**Comment by anthony-chaudhary** [comment]
Shipped the sampled-row projection floor in 9fbe0fa20 with the fail-closed production correction in aa48a406f: zero mode is invalid, F32/Q8 derive M from the selected panel, and the rectangular Q8 path matches independent-session argmax/cosine. Witness: go test -v -race ./internal/model -run '^TestLMHeadSampledRowFloor$' -count=1 passed; independent review found no aliasing, panic, or allocation defect.
